Docs for Profiling (#984)
software-mansion/protostar: website/docs/tutorials/09-profiling.md

---
sidebar_label: Profiling
---

# Profiling

## Prerequisites

> You need pprof installed beforehand. You can learn how to install it [HERE](https://github.com/google/pprof#building-pprof)

Reviewer: ```suggestion
You need pprof installed beforehand.
You can learn how to install it [in pprof documentation](https://github.com/google/pprof#building-pprof).
```

## How to profile a contract?

In case you have a test with high number of steps/memory holes, Protostar can help you diagnose which functions are the most expansive in terms of those resources.

If you want to generate profile for a test case, run:

```shell
protostar test --profiling test/test_file.cairo::test_case_name 
```
:::warning

> You can only run profiling for a single test case

Reviewer: ```suggestion
You can only run profiling for a single test case.
```

:::

Protostar will run the test in the profiling mode (it may take a little more than ususal) and produce a file `profile.pb.gz`

> Then you can read the profile using: 

Reviewer: I wonder if it would be possible to bundle pprof with Protostar 🤔

Author: I wonder as well, we can investigate it later

Author: https://github.com/software-mansion/protostar/issues/986

> ---
> sidebar_label: Profiling
> ---
> 

Reviewer: ```suggestion
```

header already does this

# Profiling

## Prerequisites

> You need pprof installed beforehand.

Reviewer 2: I think we should mention that this is only needed for interpreting the results

You can learn how to install it [in pprof documentation](https://github.com/google/pprof#building-pprof).

## How to profile a contract?

Protostar can help you diagnose which functions are the most expensive in terms of used resources.
Protostar currently measures:
- number of steps
- memory holes

> If you want to generate profile for a test case, run:

Reviewer 3: ```suggestion
If you want to generate a profile for a test case, run:
```
What is a profile?

Author: I think it is a common term, I don't think we should explain it. The same way we do not explain what terminal is.

Author: I think it is obvious from context what profile is. We have a sentence 
> Protostar can help you diagnose which functions are the most expensive in terms of used resources.

> A statistical summary of the events observed.

I do not think this is the meaning of word statistical. 

We could rephrase it to 
>Protostar can perform dynamic program analysis to help you diagnose which functions are the most expensive in terms of used resources.
 
but I am not sure if it is not unnecessary wordy. @mkaput WDYT?

```shell
protostar test --profiling test/test_file.cairo::test_case_name 
```
:::warning
You can only run profiling for a single test case.
:::

> Protostar will run the test in the profiling mode (it may take a little more than ususal) and produce a file `profile.pb.gz`

Reviewer 3: ```suggestion
Protostar will run the test in the profiling mode (it may take a little more than usual) and produce a file `profile.pb.gz`
```

Then you can read the profile using: 
```shell
go tool pprof -http=":8000" profile.pb.gz
```
### Example results

> ![Profiler](/img/prof1.jpg)

Reviewer: you can colocate images with markdown: https://docusaurus.io/docs/markdown-features/assets

> ![Profiler](/img/prof2.jpg)

Reviewer 3: [Style] This image could be scaled down to match the scaling used in `img/prof1.jpg`.

Reviewer 3: [Style] Images could be trimmed, and this document converted to mdx to control image scales from the code.

Author: I do not want to dig into using html in docs in this pr, created https://github.com/software-mansion/protostar/issues/1012

![Profiler](/img/prof3.jpg)

> 
> 
> 

Reviewer: ```suggestion
```
